planner/core/preprocess.go

```diff
@@ -564,6 +581,8 @@ func (p *preprocessor) Leave(in ast.Node) (out ast.Node, ok bool) {
 		if x.Kind == ast.BRIEKindRestore {
 			p.flag &= ^inCreateOrDropTable
 		}
+	case *ast.WithClause:
+		p.preprocessWith.inWithClause = false
```

Comment:
What happened to nested CTE, it is ok to set false if it is still in the outermost CTE?

Reply:
I only want to judge the innermost CTE.Maybe the variable name needs to be changed.  So to nested CTE,the outermost CTE will work samel at past.

Reply:
I guess this would be better.
Enter: saveInWithClause = p.preprocessWith.inWithClause
Leave: p.preprocessWith = saveInWithClause

Reply:
Sorry, can you explain in detail. The Enter and The leave is two method.How can we use the `saveInWithClause`.
I think the below var name maybe is better
`recursive -> deepestRecursive` mean the deepest CTE is recursive or not.
`inWithClause -> inDeepestWithClause` mean if is the deepest CTE WithClause or not.

Reply:
Does it work for
`with t1 as (with t11 as (select * from t)  select * from t1, t2) select * from t1;`?

Reply:
Yes.It works. And i added to `cte.test`.

Reply:
Oh.It not work. The middle `t1` did not add schema. I need think about to resolve it. I think maybe we need modify other node to record some information to distinct where is buildClause and where is the select statement behind CTE. And maybe `p.preprocessWith.name` also need to expand, to nest CTE it will record all in together now,but we should distinct CTE in different level. Can you give me some advice.Thanks

Reply:
When visiting a cte `t`.
If it's a recursive cte, save name `t` so that we can refer it in cte `t`.
If it's not a recursive cte, save name `t` when leaving this node.

Reply:
What we need to do:
1. Rewrite withName from map to a slice.
3. When Enter a withNode, set oldWithName = withName, set withNameNew = append(withName, new cte name), and record the offset
4. If is recursive, we can use withNameNew directly.
5. If is not recursive, we can use withNameNew[:len(oldWithName)+ cte_offset]
6. When leaving a withNode, set withName = append(withName, new cte name)

Reply:
The idea is that we use a stack-based method to solve it.

Reply:
But maybe the withNode have multi CTE, as below case.If we `When leaving a withNode, set withName = append(withName, new cte name)`,the `qn` will add schema.
```

SELECT
	(
		WITH qn2 AS ( SELECT 3 * a AS b FROM qn ),
                 qn AS ( SELECT 10 * a AS a FROM t1 ) SELECT
		* 
	FROM
		qn2 
		LIMIT 1 
	) 
FROM
	t1;
```

Reply:
Sorry， I don't get your point.

Reply:
When `withNode` be visited , then it will `cte.Query.Accept(v)` and it is not recursive,so `withName` and `withNameNew[:len(oldWithName)+ cte_offset]` are all empty.Beace `withNode` not reach `leave`, it is in `subExper`. When we `handleTableName` we will think `qn` is a table not a CTE. 
https://github.com/pingcap/tidb/blob/d3a02f416aaa1eda89ce6bec2242ea8254ccfa46/parser/ast/dml.go#L1143-L1156

Reply:
`it is not recursive,so withName and withNameNew[:len(oldWithName)+ cte_offset] are all empty`

When Enter a withNode, set oldWithName = withName, set withNameNew = append(withName, new cte name), and record the offset.
It's not empty.

Reply:
withName: []
withNameNew: [qn2, qn].
When processing qn2, it can see [qn2], so it would report an error.

Reply:
I think that would be easier:

```
func (n *WithClause) Accept(v Visitor) (Node, bool) {
	newNode, skipChildren := v.Enter(n)
	if skipChildren {
		return v.Leave(newNode)
	}

	for _, cte := range n.CTEs {
		**v.Enter(cte)**
		node, ok := cte.Query.Accept(v)
		if !ok {
			return n, false
		}
		cte.Query = node.(*SubqueryExpr)
		**v.Leave(cte)**
	}
	return v.Leave(n)
}
```

Reply:
>If is recursive, we can use withNameNew directly.
If is not recursive, we can use withNameNew[:len(oldWithName)+ cte_offset]

I think when we firstly deal with `withNode`. And withNameNew:[qn2,qn],if it is recursive,we can use [qn2,qn],but if it is not recursive, i think it should be [].Otherwise it doesn't look any different

>When processing qn2, it can see [qn2], so it would report an error.

If qn2 is really a table,it should not be report an error.It need dealt as `table`
`WITH qn2 AS ( SELECT 3 * a AS b FROM qn2)` -> ``WITH qn2 AS ( SELECT 3 * a AS b FROM `test`.`qn2` )``

Reply:
> I think that would be easier:
> 
> ```
> func (n *WithClause) Accept(v Visitor) (Node, bool) {
> 	newNode, skipChildren := v.Enter(n)
> 	if skipChildren {
> 		return v.Leave(newNode)
> 	}
> 
> 	for _, cte := range n.CTEs {
> 		**v.Enter(cte)**
> 		node, ok := cte.Query.Accept(v)
> 		if !ok {
> 			return n, false
> 		}
> 		cte.Query = node.(*SubqueryExpr)
> 		**v.Leave(cte)**
> 	}
> 	return v.Leave(n)
> }
> ```
It seem like a good idea.But we need care about Iterator problem.if as below order, `qn` will be missed
```
1.qn2 AS ( SELECT 3 * a AS b FROM qn ),
2.qn AS ( SELECT 10 * a AS a FROM t1 )
```

Reply:
It's ok, qn2 couldn't see qn, it should return an error for this case.

Reply:
This case is found by `cte.test` , i see it can work. What is the error you say,i think i miss something
https://github.com/pingcap/tidb/blob/d3a02f416aaa1eda89ce6bec2242ea8254ccfa46/cmd/explaintest/t/cte.test#L128-L132

Reply:
The order is different.